src/background/providers/llama.ts

```diff
@@ -0,0 +1,63 @@
+import { fetchSSE } from '../fetch-sse'
+import { GenerateAnswerParams, Provider } from '../types'
+
+export class LLAMAProvider implements Provider {
+  constructor(private token: string, private model: string) {
+    this.token = token
+    this.model = model
+  }
+
+  private buildPrompt(prompt: string): string {
+    return '"' + prompt + '"'
+  }
+
+  async generateAnswer(params: GenerateAnswerParams) {
+    let result = ''
+    await fetchSSE('https://api.together.xyz/inference', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${this.token}`,
+      },
+      body: JSON.stringify({
+        model: `${this.model}`,
+        max_tokens: 2048,
+        prompt: this.buildPrompt(params.prompt),
+        request_type: 'language-model-inference',
+        temperature: 0.7,
+        top_p: 0.7,
+        top_k: 50,
+        repetition_penalty: 1,
+        stream_tokens: true,
+      }),
+      onMessage(message: string) {
+        if (message === '[DONE]') {
+          params.onEvent({ type: 'done' })
+          return
+        }
+
+        let data
+        try {
+          data = JSON.parse(message)
+        } catch (err) {
+          console.error(err)
+          return
+        }
+
+        const text = data.choices?.[0]?.text
+        if (text) {
+          result += text
```

Comment:
Interesting ❓ I expect the caller site or even receiver would add a new text snippets to the end of existing UI or buffer or something. The code continuously posts the whole text at every new text from the server. 

If the current approach works seamlessly, i'm fine with the pr. Otherwise please revise it. 

In advance I'm going to approve the pr because I do not want to waste time due to the time zone difference. Please take my concern above. Otherwise everything look awesome!

Reply:
I think here the stream can be stopped by '[DONE]' signal, if you are worried about this.
<img width="359" alt="Screen Shot 2023-08-24 at 23 39 13" src="https://github.com/hunkimForks/chatgpt-arxiv-extension/assets/13414571/b011e7f9-77ee-4d74-a62d-69bb4f114fc4">

Reply:
The 'DONE' handling part is ok. 

What I concern is the content of event raised through `params.onEvent` when a text snippet arrives from a LLM backend via the streaming. 

Let's say the LLM on the other end streams a text sequence "HELLO" -> " WORLD" -> "!". According to the current implementation, `params.onEvent` gets called three times with contents as following order:
1. `HELLO`
2. `HELLO WORLD`
3. `HELLO WORLD!`

My expectation here is that the `HELLO` should only get raised when it hasn't been fired before va `params.onEvent`. Not sure the handler of `params.onEvent` processes such a sequence that allows duplications. If the caller expects the whole text received everytime a new word or phrases available, the implementation looks good. Otherwise, I think it should be changed in a way those words in text streams should be propagated to the next handler once.